### ISSUE TITLE
Godot 4.0 Terrain tiles not updating

### ISSUE THREAD

**Issue by mageowl** [godotengine/godot#64674]
### Godot version

4.0.alpha14

### System information

MacOS 12.5, 2.3 GHz Dual-Core Intel Core i5

### Issue description

The terrain tile placement in Godot 4 is often bugged, and will end up with tiles I didn't place, or tiles not updating after placing new ones

https://user-images.githubusercontent.com/32573897/185769867-32bde2b3-d81d-4d19-bd63-c7f122f60c82.mov

.

### Steps to reproduce

1. Create a minimal tileset setup with terrain.
2. Draw an odd shape.
3. About 30% of the time it will be broken.

### Minimal reproduction project

[Archive.zip](https://github.com/godotengine/godot/files/9388184/Archive.zip)

**Comment by groud** [comment]
I think this is more or less a wontfix.

The algorithm is now quite complex, it tries to find a good working solution for each situation, trying to minimize how many terrain bits or transitions gets modified. When your terrain does not have a lot of possible combinations (as yours does), then it may not find a tile to fit in there . It may modify the surrounding tiles, but to avoid testing all possible combinations it limits itself to the first surrounding tiles, and does some sort of a guess to find a solution. We can't really make this more complex however, as it would slow down significantly the painting process.

We may try to change the constraints weights here and there in the algorithm to see if it improves the situation, but I don't think it will be possible to improve the situation that much.

**Comment by mageowl** [comment]
Ok, thanks for saying. Yeah, I guess I haven't tried it with a full tile set, as I usually don't need one.
:)

**Comment by bitbrain** [comment]
>  We can't really make this more complex, however, as it would slow down significantly the painting process.

Tiled has a similar issue. To solve this, they offer a button to "hard refresh" the entire tilemap to ensure the rendering is displayed correctly.

**Comment by Shadetheartist** [comment]
I appreciate the complexity of tile painting, and i have looked at the Tilemap source code. Though i'm only a few hours deep into the problem i might be able to help with a solution.

I have developed a dynamic tile painting algorithm in the past for a Carcassonne (board game) simulator written in go. If i remember correctly it would compare tiles based on an internal 3x3 'corners and edges' matrix which is analogous to how terrain painting is done in godot. The algorithm was incredibly lightweight, and could compare hundreds of tiles together in microseconds. 

![image](https://user-images.githubusercontent.com/8973930/209749076-5779e66e-e57b-4185-9747-9a783463ed13.png)

I might take a shot at improving this, as the terrain issue stands directly in the way of progress in my motherload-esque game idea. My actual problem is something in the same vein as this issue. In the editor this tool actually works perfectly, but when i try to replicate the paint and erase tool in gdscript, i have only `set_cells_terrain_connect` and `set_cells_terrain_path` as the API to the terrain tooling, and after trying various approaches i cannot get it to function the same as the editor. Maybe someone involved in writing the GDscript API could advise on how it's intended to be used when trying to erase cells?

(using editor - works well)

![image](https://user-images.githubusercontent.com/8973930/209750602-eb5e2d80-c24d-4412-b60a-cf4249d76a5e.png)

(one attempt of _many_ to get the GDscript API to work like the editor)

![image](https://user-images.githubusercontent.com/8973930/209750773-3ff6c768-2e77-4cf0-a29d-b4f3edc99a53.png)

**Comment by furroy** [comment]
I too struggled with getting the same results with script calls to set_cells_terrain_connect() to work the same as the editor. I had to pass in false for the last optional parameter dealing with empty cells which helped a lot. But still the tiles don't update/repaint 100% of the time.

**Comment by andypandy47** [comment]
Seeing this issue as well. The top set of blocks was connected using a call to set_cells_terrain_connect() and the bottom was painted in the editor. Their collision shapes should be identical.
<img width="433" alt="image" src="https://user-images.githubusercontent.com/19389664/222964526-f5f47996-b15e-4d9a-bd00-b2046386e46d.png">

**Comment by Janonas** [comment]
Im having issues with set_cells_terrain_connect(), especially if im using it to patch the terain after removing some tiles in game. Its quite ruining my project.

**Comment by furroy** [comment]
Ofc, don't forget to call update_bitmask_region() after changing cells to rebuild autotiles. If you think this might be the issue, try calling with no args and if it does fix it, you can calc more specific area to update for better performance.

**Comment by Janonas** [comment]
Ah, im using godot 4.0, where no such equivalent function seems to exist.

**Comment by andypandy47** [comment]
I've been playing around with procedural generation in Godot4 and have found the terrain system to be somewhat inconsistent when calling set_cells_terrain_connect. I frequently get instances where chunks have generated but the terrain has bugged out and can't quite figure out why.
<img width="888" alt="Screenshot 2023-03-25 100219" src="https://user-images.githubusercontent.com/19389664/227710626-d83b5202-dfe6-411d-bdf1-eb7d46d61352.png">
<img width="593" alt="messed_up_terrain" src="https://user-images.githubusercontent.com/19389664/227710637-6828ad05-d918-4834-a3a9-7b507009cdb9.png">

In those screenshots you can see that the terrain has been generated as the colliders for the edge tiles have been placed yet no actual sprite is drawn for some reason. Whats annoying is that if I move the player far enough away and come back to regenrate the terrain it works fine.

It's possible this could be caused by something in the way I'm generating the terrain but it works fine throughout the project just every now and again a little patch like in the screenshots above doesn't seem to generate properly.

**Comment by andypandy47** [comment]
I think my issue is down to using Task.Run in C# to help avoid lag spikes when generating terrain. Will raise a separate issue for this as it seems to be happening quite consistently

**Comment by Salvakiya** [comment]
@groud has done some work with the tilemaps before. Could you maybe help direct where I should look to fix this? looking to make a hotfix.

(specifically referring to this https://github.com/godotengine/godot/pull/67390)

**Comment by mageowl** [comment]
![Screenshot 2023-07-01 at 4 49 36 PM](https://github.com/godotengine/godot/assets/32573897/be4474a4-1a4f-4234-a562-edef9ae094b2)

Even when there is a correct tile for the autotile, with 2 different sets of terrain, Godot chooses incorrectly.

**Comment by mageowl** [comment]
(desired solution:
![Screenshot 2023-07-01 at 4 52 27 PM](https://github.com/godotengine/godot/assets/32573897/32d7dbf9-5d9b-4956-82b5-4f8e05e869a5)
)

**Comment by filipinyo** [comment]
What's the situation with this? Seems like a big stopper for a procedurally generated games :/

**Comment by krazyjakee** [comment]
> ![Screenshot 2023-07-01 at 4 49 36 PM](https://github.com/godotengine/godot/assets/32573897/be4474a4-1a4f-4234-a562-edef9ae094b2)
> 
> Even when there is a correct tile for the autotile, with 2 different sets of terrain, Godot chooses incorrectly.

@groud correct me if I'm wrong but isn't this the proof that this is a legitimate bug and not a "won'tfix"?

**Comment by groud** [comment]
> @groud correct me if I'm wrong but isn't this the proof that this is a legitimate bug and not a "won'tfix"?

No. A single picture without an MRP isn't enough to explain it as a bug, and I explained what were the tradeoff of the algorithm implementation with my first comment.

That being said, I am not against a rework of the terrain system as a whole that would fix the original issue, I just don't have the time to work on it. It also requires a solid proposal, that would understand all the tradeoffs each solutions will require.
 
In the meantime, you can try this plugin: https://github.com/Portponky/better-terrain, it seems it had made several people happy (it has limitations too, but they might now apply to you project)

**Comment by KoBeWi** [comment]
There are a couple of proposals, but idk if they count as solid:
https://github.com/godotengine/godot-proposals/issues/7670
https://github.com/godotengine/godot-proposals/issues/5575

**Comment by hsandt** [comment]
v4.6.stable.official [89cea1439] update: I have slightly different issue than OP, where rectangle painting in particular fails, pencil painting covers 75% of the cases but misses a corner (and in fact before taking the video, it also failed on edges), but deleting empty space around the tiles forces refresh indeed.

Video:

https://github.com/user-attachments/assets/7281c9ef-7225-4908-8894-366a4cbfcded

MWE (not a full project, just a scene + tileset + tile sheet to import in a project):
[Godot 4.0 Terrain tiles not updating MWE.zip](https://github.com/user-attachments/files/25727647/Godot.4.0.Terrain.tiles.not.updating.MWE.zip)

EDIT 2026-06-24: issue occurred again on another project with more complex tiles (some tiles have multiple variations), rectangle tool will only shuffle random tiles while keeping the bad connections (which means it _does_ edit the tiles but still doesn't read constraints properly).

I also tested the Line tool and even the Bucket Tool, they have the same issue.
